cardano-cli/src/Cardano/CLI/Read.hs

```diff
@@ -1086,13 +1099,38 @@ instance Error CostModelsError where
         ]
 
 readCostModels
-  :: File L.CostModels In
+  :: ()
+  => ShelleyBasedEra era
+  -> File L.CostModels In
   -> ExceptT CostModelsError IO L.CostModels
-readCostModels (File fp) = do
+readCostModels sbe (File fp) = do
   bytes <- handleIOExceptT (CostModelsErrorReadFile . FileIOError fp) $ LBS.readFile fp
   costModels <- firstExceptT (CostModelsErrorJSONDecode fp) . except $ Aeson.eitherDecode bytes
   when (null $ fromAlonzoCostModels costModels) $ throwE $ CostModelsErrorEmpty fp
+  forM_ (allValues @L.Language) $ checkCostModelSize costModels
   return costModels
+ where
+  checkCostModelSize :: L.CostModels -> L.Language -> ExceptT CostModelsError IO ()
+  checkCostModelSize models lang =
```

Comment:
I think those checks should be in `cardano-api` so we can reuse them in decoding genesis as well, because genesis can contain PV2 and PV3 models. This is used only in starting testnets afaik.